Add parsing support for `background-clip: border-area`

smfr opened the pull request:
#### 7f71d31815a84515504ba6b38e379e2a1a6f0a44
<pre>
Add parsing support for `background-clip: border-area`
<a href="https://bugs.webkit.org/show_bug.cgi?id=277842">https://bugs.webkit.org/show_bug.cgi?id=277842</a>
<a href="https://rdar.apple.com/133503570">rdar://133503570</a>

Reviewed by Tim Nguyen.

Add parsing support for the new `border-area` value in `background-clip`[1]. There&apos;s a new keyword
in CSSValueKeywords.in, and a new value to the `FillBox` enum, and the relevant conversions.

Sadly we can&apos;t use generated code for parsing any more, because `background-clip` uses
a sub-grammar which doesn&apos;t support conditional keywords. I tried to fix the Python code
generation, but it&apos;s a non-trivial task.

Flip the setting to testable, and modify relevant WPT to test for this new value.

[1] <a href="https://drafts.csswg.org/css-backgrounds-4/#background-clip">https://drafts.csswg.org/css-backgrounds-4/#background-clip</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-computed.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeBackgroundComponent):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundClip):
(WebCore::CSSPropertyParserHelpers::consumeBackgroundClip):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: <a href="https://commits.webkit.org/282049@main">https://commits.webkit.org/282049@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02077eaf23773b1d9c30dc5efd56a5a1460ece07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61825 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14417 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65804 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12370 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48865 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12641 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49829 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8567 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38240 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53546 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30661 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34885 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10779 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11301 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54920 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56694 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67533 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61066 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5768 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10819 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57210 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5793 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57446 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13762 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4728 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82829 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36979 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14494 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38063 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39159 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37808 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-computed-expected.txt:
```diff
@@ -2,5 +2,6 @@
 PASS Property background-clip value 'content-box'
-PASS Property background-clip value 'border-box, padding-box, content-box'
+PASS Property background-clip value 'border-area'
+PASS Property background-clip value 'border-box, padding-box, content-box' border-area
```
This baseline seems suspicious

smfr committed:
Add parsing support for `background-clip: border-area`
https://bugs.webkit.org/show_bug.cgi?id=277842
rdar://133503570

Reviewed by Tim Nguyen.

Add parsing support for the new `border-area` value in `background-clip`[1]. There's a new keyword
in CSSValueKeywords.in, and a new value to the `FillBox` enum, and the relevant conversions.

Sadly we can't use generated code for parsing any more, because `background-clip` uses
a sub-grammar which doesn't support conditional keywords. I tried to fix the Python code
generation, but it's a non-trivial task.

Flip the setting to testable, and modify relevant WPT to test for this new value.

[1] https://drafts.csswg.org/css-backgrounds-4/#background-clip

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-clip-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-computed.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeBackgroundComponent):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundClip):
(WebCore::CSSPropertyParserHelpers::consumeBackgroundClip):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: https://commits.webkit.org/282049@main
